danny-avila/LibreChat
# test/playwright jest cicd

This PR makes changes to help streamline the development process by providing an automated authentication process for playwright, separating out configuration for local vs. pipeline runs, and setting up CI/CD pipelines for unit and integration tests with github workflows.

- [x] Create automated authentication process for playwright
- [x] Add new secrets to github actions settings on fork (will need to be done for this repo by @danny-avila)
- [x] Update e2e run scripts and remove e2e:auth and e2e:auth-test scripts
- [x] Create separate playwright configs for local and pipeline environments
- [x] Update github action for playwright 
- [x] Update Jest config for frontend and create pipeline to run unit tests on each PR
- [x] Update test script in root package.json to run clientside unit tests from root folder
- [x] Move backend jest config to /api
- [x] Create action pipeline for backend unit test runs

Current Issues:
- Playwright pipeline: While the playwright run with authentication works fine locally, It still has an issue in the pipeline where the UI does not seem to load properly. I have yet to determine why this happens, however getting it to install, build, and run so it could get to the point of running the tests was a significant feat.
- API test pipeline: There is an issue with the api jest pipeline where it is unable to install the Sharp library. I tried the suggestion provided by the error by adding a step that manually installs the linux x64 distro of Sharp but it didn't make any difference. Once this is resolved I expect the pipeline to run fine.

## Type of change


## How Has This Been Tested?
- Ran tests locally, confirmed to be working as expected
- Ran pipelines against my branch on fork


### **Test Configuration**:
- Add E2E_USER_EMAIL and E2E_USER_PASSWORD to .env.test in /e2e folder (NOTE: this user must exist in the mongodb you are testing against)
- Update project and run `npm run e2e:debug`, see that it automatically authenticates. 
- Run `npm run test:client` from root folder, see that the unit tests pass as expected.
- Run `npm run test:api` from root
- cd /client and `npm run test:ci` to test the ci script
- cd /api and `npm run test:ci` to test the api ci script (note that tests will fail since they get the env vars from github secrets)

To get the github action working, the following should be added to the github action secrets (this should be done before merging the PR): 
- `E2E_USER_EMAIL` and `E2E_USER_PASSWORD`: this should be a valid user in the db that the MONGO_URI is connecting to
- `CREDS_KEY` and `CREDS_IV`
- JWT_SECRET

*Note: you may opt to remove lines 4 and 5 of the pipelines as it may be unnecessary to also run unit tests when merging in addition to PRs.*

## Checklist:
